From 08235d93f9969750de1b6c5d5794da6cc7f63046 Mon Sep 17 00:00:00 2001
From: Tao Huang <huangtao@rock-chips.com>
Date: Mon, 16 Dec 2019 19:31:34 +0800
Subject: [PATCH] soc: rockchip: sdmmc_vendor_storage: Add g_vendor free

Free g_vendor when init failed or module exit.
Alloc g_vendor once because g_vendor size is large.

Change-Id: Ia4909ba86f010c59fc2821dd0732b698b188bbcc
Signed-off-by: Tao Huang <huangtao@rock-chips.com>
---
 drivers/soc/rockchip/sdmmc_vendor_storage.c | 18 +++++++++++-------
 1 file changed, 11 insertions(+), 7 deletions(-)

diff --git a/drivers/soc/rockchip/sdmmc_vendor_storage.c b/drivers/soc/rockchip/sdmmc_vendor_storage.c
index f778e67a4c1b..55683e3a1697 100644
--- a/drivers/soc/rockchip/sdmmc_vendor_storage.c
+++ b/drivers/soc/rockchip/sdmmc_vendor_storage.c
@@ -93,10 +93,6 @@ static int emmc_vendor_storage_init(void)
 	u32 i, max_ver, max_index;
 	u8 *p_buf;
 
-	g_vendor = kmalloc(sizeof(*g_vendor), GFP_KERNEL | GFP_DMA);
-	if (!g_vendor)
-		return -ENOMEM;
-
 	max_ver = 0;
 	max_index = 0;
 	for (i = 0; i < EMMC_VENDOR_PART_NUM; i++) {
@@ -135,8 +131,6 @@ static int emmc_vendor_storage_init(void)
 	}
 	return 0;
 error_exit:
-	kfree(g_vendor);
-	g_vendor = NULL;
 	return -1;
 }
 
@@ -597,6 +591,10 @@ static int vendor_init_thread(void *arg)
 {
 	int ret, try_count = 5;
 
+	g_vendor = kmalloc(sizeof(*g_vendor), GFP_KERNEL | GFP_DMA);
+	if (!g_vendor)
+		return -ENOMEM;
+
 	do {
 		ret = emmc_vendor_storage_init();
 		if (!ret) {
@@ -609,6 +607,9 @@ static int vendor_init_thread(void *arg)
 	if (!ret) {
 		ret = misc_register(&vender_storage_dev);
 		rk_vendor_register(emmc_vendor_read, emmc_vendor_write);
+	} else {
+		kfree(g_vendor);
+		g_vendor = NULL;
 	}
 	pr_info("vendor storage:20190527 ret = %d\n", ret);
 	return ret;
@@ -623,8 +624,11 @@ static int __init vendor_storage_init(void)
 
 static __exit void vendor_storage_deinit(void)
 {
-	if (g_vendor)
+	if (g_vendor) {
 		misc_deregister(&vender_storage_dev);
+		kfree(g_vendor);
+		g_vendor = NULL;
+	}
 }
 
 device_initcall_sync(vendor_storage_init);
-- 
2.35.3

